src/tuner/nccl_ofi_model.c

```diff
@@ -74,167 +66,208 @@ static const float nccl_base_lat[NCCL_NUM_ALGORITHMS][NCCL_NUM_PROTOCOLS] = {
-		const double bandwidth = NCCL_OFI_TUNER_INTERNODE_BW * per_proto_bandwidth_overhead;
-		const double net_latency = NCCL_OFI_TUNER_COMPLETION_OVERHEAD + NCCL_OFI_TUNER_NETWORK_LATENCY;
-		latency = (p2p_latency * num_steps_intranode +
-			   num_steps_internode * (net_latency + base_latency));
-		return latency + (num_trees_per_rank * per_tree_msg_size) / bandwidth;
+		/* Latency overheads, no bandwidth overheads. */
+		internode_latency_per_step += NCCL_OFI_TUNER_COMPLETION_OVERHEAD;
+		break;
 	}
 	default: {
 		return INFINITY;
 	}
 	}
+
+	const double transfer_size = messageSize / num_trees_per_host;
+	const double intranode_latency_cost = intranode_latency_per_step * num_critical_steps_intranode;
+	const double internode_latency_cost = internode_latency_per_step * num_critical_steps_internode;
+
+	const double internode_bandwidth_cost = num_total_steps_internode * transfer_size / internode_bandwidth;
+	const double intranode_bandwidth_cost = num_total_steps_intranode * transfer_size / intranode_bandwidth;
+
+	return internode_latency_cost + intranode_latency_cost + intranode_bandwidth_cost + internode_bandwidth_cost;
 }
 
 double all_reduce_nvlstree_cost(struct nccl_ofi_tuner_model_dims const* dims,
 				int proto,
 				int pipe_ops,
+				const size_t chan,
 				const size_t nBytes)
 {
-	const double bandwidth = (NCCL_OFI_TUNER_INTERNODE_BW + NCCL_OFI_TUNER_INTRANODE_BW);
+	if (proto != NCCL_PROTO_SIMPLE)
+		return INFINITY;
+
+	const double ranks_per_node = ceil(dims->num_ranks * 1.0 / dims->num_nodes);
+	if (chan < ranks_per_node)
+		return INFINITY;
+
 	const double p2p_latency = nccl_nvlink_lat[NCCL_ALGO_NVLS_TREE][proto];
 	const double base_latency = nccl_base_lat[NCCL_ALGO_NVLS_TREE][proto];
-	const double num_steps_internode = 2 * log(dims->num_nodes) / log(2) - 1;
-	const double num_steps_intranode = 2 * log(dims->num_nodes) / log(2) - 1;
-	const double intranode_latency = p2p_latency * num_steps_intranode;
-	double per_tree_msg_size = 2 * nBytes;
 
-	switch (proto) {
-	case NCCL_PROTO_LL: {
-		return INFINITY;
-	}
-	case NCCL_PROTO_LL128: {
-		return INFINITY;
-	}
-	case NCCL_PROTO_SIMPLE: {
-		const double net_latency = NCCL_OFI_TUNER_COMPLETION_OVERHEAD + NCCL_OFI_TUNER_NETWORK_LATENCY;
-		const double internode_latency = (net_latency + base_latency) * num_steps_internode;
-		double latency = (intranode_latency + internode_latency);
-		return latency + (2 * per_tree_msg_size) / bandwidth;
-	}
-	default: {
-		return INFINITY;
-	}
-	}
+	const double num_trees_per_host = 2;
+	const double tree_height = floor(log(NCCL_OFI_IS_POWER_OF_TWO(dims->num_nodes) ? dims->num_nodes : dims->num_nodes - 1) / log(2)) + 1;
+
+	const double num_critical_steps_internode = /* there is a critical internode set of transfers */
+		2 * tree_height /* on the way up the tree, and again on the way back down. */
+		;
+
+	const double num_critical_steps_intranode = /* there is a critical-path intranode nvswitch process */
```

Comment:
this seems `ranks_per_node * num_trees_per_host` is taken into the consideration here because of possible overhead to latency calculate when ranks_per_node increases. I suggest we keep this variable simple as just `2` without `ranks_per_node * num_trees_per_host` ,  and we add `ranks_per_node * num_trees_per_host` overhead into latency. See below comment for this.

Reply:
Same as other threads:

4164ff7 fixup! tuner: better ring rank/msize binning